File: PerDayOneQuestion/367_有效的完全平方数.go
/*
-------------------------------------
# @Time    : 2021/11/4 0:23:18
# @Author  : Giyn
# @Email   : giyn.jy@gmail.com
# @File    : 367_有效的完全平方数.go
# @Software: GoLand
-------------------------------------
*/

package main

import "fmt"

func main() {
	num := 104976
	fmt.Println(isPerfectSquare(num))
}

func isPerfectSquare(num int) bool {
	if num == 1 {
		return true
	}
	left, right := 1, num
	for left <= right {
		mid := left + ((right - left) >> 1)
		if mid*mid == num {
			return true
		} else if mid*mid < num {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return false
}

File: array/367_有效的完全平方数/367_有效的完全平方数.py
"""
-------------------------------------
# -*- coding: utf-8 -*-
# @Time    : 2021/9/30 23:08:20
# @Author  : Giyn
# @Email   : giyn.jy@gmail.com
# @File    : 367_有效的完全平方数.py
# @Software: PyCharm
-------------------------------------
"""


class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        if num == 1:
            return True

        left = 1
        right = num

        while left <= right:
            mid = left + ((right - left) >> 1)
            if mid ** 2 == num:
                return True
            elif mid ** 2 > num:
                right = mid - 1
            else:
                left = mid + 1

        return False


if __name__ == '__main__':
    solution = Solution()
    print(solution.isPerfectSquare(16))
    print(solution.isPerfectSquare(14))
